include/gridtools/common/hugepage_alloc.hpp

```diff
@@ -11,25 +11,202 @@
+                    MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | MAP_NORESERVE,
+                    -1,
+                    0);
+                if (ptr == MAP_FAILED)
+                    throw std::bad_alloc();
+                break;
+            }
+            return {ptr, size};
+        }
+
+        inline void deallocate(void *ptr, std::size_t size, hugepage_mode mode) {
+            switch (mode) {
+            case hugepage_mode::disabled:
+                free(ptr);
+                break;
+            default:
+                if (munmap(ptr, size))
+                    throw std::bad_alloc();
+                break;
+            }
+        }
+#else
+        inline std::size_t cache_line_size() {
+            return 64; // default value for x86-64 archs
+        }
+
+        inline std::size_t cache_sets() {
+            return 64; // default value for (most?) x86-64 archs
+        }
+
+        inline std::size_t hugepage_size() {
+            return 2 * 1024 * 1024; // 2MB is the default on most systems
+        }
+
+        inline std::size_t page_size() {
+            return 4 * 1024; // 4kB is the default on most systems
+        }
+
+        inline std::tuple<void *, std::size_t> allocate(std::size_t size, hugepage_mode) {
+            // here we hope that aligning to hugepage_size will return transparent huge pages
+            void *ptr;
+            size = ((size + hugepage_size() - 1) / hugepage_size()) * hugepage_size();
+            if (posix_memalign(&ptr, hugepage_size, size))
+                throw std::bad_alloc();
+            return {ptr, size};
+        }
+
+        inline void deallocate(void *ptr, std::size_t, hugepage_mode) { free(ptr); }
+#endif
+
+        inline std::size_t allocation_offset() {
+            static std::atomic<std::size_t> s_offset(0);
+            return ((s_offset++ % cache_sets()) << ilog2(cache_line_size())) + cache_line_size();
+        }
+
+        inline hugepage_mode hugepage_mode_from_env() {
+            const char *env_value = std::getenv("GT_HUGEPAGE_MODE");
+            if (!env_value || std::strcmp(env_value, "transparent") == 0)
+                return hugepage_mode::transparent;
+            if (std::strcmp(env_value, "disable") == 0)
+                return hugepage_mode::disabled;
+            if (std::strcmp(env_value, "explicit") == 0)
+                return hugepage_mode::explicit_allocation;
+            std::fprintf(stderr, "warning: env variable GT_HUGEPAGE_MODE set to invalid value '%s'\n", env_value);
+            return hugepage_mode::transparent;
+        }
+
+        struct ptr_metadata {
+            std::size_t offset, full_size;
+            hugepage_mode mode;
+        };
+
+    } // namespace hugepage_alloc_impl_
 
     /**
      * @brief Allocates huge page memory (if GT_NO_HUGETLB is not defined) and shifts allocations by some bytes to
      * reduce cache set conflicts.
      */
     inline void *hugepage_alloc(std::size_t size) {
```

Comment:
I doubt that it still makes sense to keep all functions `inline`.
How about converting it to a real library? The only exported functions are `hugepage_alloc` and `hugepage_free`, aren't they? I would export them in `c`-style -- with `extern "C"` linkage, guarded with `#ifdef __cplusplus__`.

Reply:
Clearly. But it would also change GridTools from being header-only to not being header-only, right? I was not sure if somebody could possibly be bothered by this?

Reply:
Frankly, I don't see any complications with that. I do know where the fears come from:
  - in the past we had a single `gridtools` target that we offered to the user -- the decision on being `header only` was `viral`.
  - both headers and cpp files were depended on user defined macros -- it was extremely hard to maintain consistency between the headers and the library binary.

The current case is very different:
  - we export the collection of targets that form the dependency structure. Only two targets will have dependency on `hugepage_alloc` -- `stencil_cpu_ifirst` and `storage_cpu_ifirst`. They will still be `header only` libraries with `interface` (in cmake sense) dependency on real library.
 -  `hugepage_alloc` is not driven by user defined macros.

An extra durability will come from the fact that all exported functions have `c` linkage -- being compiled and installed with one compiler,  `hugepage_alloc` binary could be linked using any other compiler.

Reply:
No. At this point it is not possible to introduce a compiled library in our CMake without too much effort.

Reply:
So should we keep it inline, or what’s the best strategy?